README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
kader
=====

The goal of kader is to supply functions to compute nonparametric kernel estimators for

-   density estimation using a data-adjusted kernel or an appropriate rank-transformation, and for
-   regression using a data-adjusted kernel.

The functions are based on the theory introduced in

-   Srihera, R., Stute, W. (2011): Kernel adjusted density estimation. Statistics and Probability Letters 81, 571 - 579, URL <http://dx.doi.org/10.1016/j.spl.2011.01.013>.
-   Eichner, G., Stute, W. (2012): Kernel adjusted nonparametric regression. Journal of Statistical Planning and Inference 142, 2537 - 2544, URL <http://dx.doi.org/10.1016/j.jspi.2012.03.011>.
-   Eichner, G., Stute, W. (2013): Rank Transformations in Kernel Density Estimation. Journal of Nonparametric Statistics 25(2), 427 - 445, URL <http://dx.doi.org/10.1080/10485252.2012.760737>.

A very brief summary of the theory and sort of a vignette is presented in Eichner, G. (2017): Kader - An R package for nonparametric kernel adjusted density estimation and regression. In: Ferger, D., et al. (eds.): From Statistics to Mathematical Finance, Festschrift in Honour of Winfried Stute. Springer International Publishing. To appear in Oct. 2017.

Installation
------------

You can install kader from CRAN with:

``` r
install.packages("kader")
```

or from github with:

``` r
# install.packages("devtools")
devtools::install_github("GerritEichner/kader")
```

Example
-------

This example shows you how to estimate at *x*<sub>0</sub> = 2 the value of the density function of the probability distribution underlying Old-Faithful's eruptions data using the (nonrobust) method of Srihera & Stute (2011). The initial grid (given to `Sigma`) on which the minimization of the estimated MSE as a function of a (kernel-adjusting) scale parameter *σ* is started is rather coarse here to save computing time.

``` r
library(kader)
x0 <- 2
sigma <- seq(0.01, 10, length = 21)
fit <- kade(x = x0, data = faithful$eruptions, method = "nonrobust",
  Sigma = sigma, ticker = TRUE)
#> h set to n^(-1/5) with n = 272.
#> theta set to arithmetic mean of data in faithful$eruptions.
#> Using the adaptive method of Srihera & Stute (2011)
#> For each element in x: Computing estimated values of
#> bias and scaled variance on the sigma-grid.
#> Note: x has 1 element(s) and the sigma-grid 21.
#> 
#> As a little distraction, the 'ticker' documents the
#> computational progress (if you have set ticker = TRUE).
#> x[1]:sigma:1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21.
#> Minimizing MSEHat:
#> Step 1: Search smallest maximizer of VarHat.scaled on sigma-grid.
#> Step 2: Search smallest minimizer of MSEHat on sigma-grid to the
#>         LEFT of just found smallest maximizer of VarHat.scaled.
#> Step 3: Finer search for 'true' minimum of MSEHat using
#>         numerical minimization. (May take a while.)
#> sigma:1.sigma:1.sigma:1.sigma:1.sigma:1.sigma:1.sigma:1.sigma:1.sigma:1.sigma:1.sigma:1.sigma:1.sigma:1.
#> Step 4: Check if numerically determined minimum is smaller
#>         than discrete one.
#>         Yes, optimize() was 'better' than grid search.
#> 
#> .
print(fit)
#>   x         y sigma.adap  msehat.min discr.min.smaller sig.range.adj
#> 1 2 0.5478784   1.996629 0.003822793             FALSE             0
```


NEWS.md
# kader 0.0.8

* First release of kader.


R/kader.R
#2345678901234567890123456789012345678901234567890123456789012345678901234567890
#' Kernel Adjusted Density Estimation and Regression
#'
#' Package of functions to compute kernel estimators for
#' \itemize{
#'  \item nonparametric density estimation using a data-adjusted kernel or an
#'        appropriate rank-transformation, and for
#'  \item nonparametric regression using a data-adjusted kernel.
#' }
#'
#' The functions are based on the theory laid out in the following papers:
#' \itemize{
#'  \item Srihera, R., Stute, W. (2011): Kernel adjusted density
#'        estimation. Statistics and Probability Letters 81, 571 - 579,
#'        URL \url{http://dx.doi.org/10.1016/j.spl.2011.01.013}.
#'  \item Eichner, G., Stute, W. (2012): Kernel adjusted nonparametric
#'        regression. Journal of Statistical Planning and Inference 142,
#'        2537 - 2544, URL \url{http://dx.doi.org/10.1016/j.jspi.2012.03.011}.
#'  \item Eichner, G., Stute, W. (2013): Rank Transformations in Kernel
#'        Density Estimation. Journal of Nonparametric Statistics 25(2),
#'        427 - 445, URL \url{http://dx.doi.org/10.1080/10485252.2012.760737}.
#' %  \item Eichner, G., Stute, W. (2015): Rank-Based Kernel Smoothing --
#' %        \eqn{L_2}-approach. Talk presented at the 12th Workshop on
#' %        Stochastic Models, Statistics and Their Applications in Wroclaw,
#' %        Poland, February 2015.
#' }
#'
#' A very brief summary of the three papers above and sort of a vignette is
#' presented in Eichner, G. (2017): Kader - An R package for nonparametric
#' kernel adjusted density estimation and regression. In: Ferger, D., et al.
#' (eds.): From Statistics to Mathematical Finance, Festschrift in Honour of
#' Winfried Stute. Springer International Publishing. To appear in Jan. 2018.
#' DOI then(!) presumably: 10.1007/978-3-319-50986-0.
#'
#' @docType package
#' @name kader
NULL


R/kernels.R
#2345678901234567890123456789012345678901234567890123456789012345678901234567890
### ~/PapersBooksReportsReviewsTheses/PapersSelbst/FestschriftWS70/kader/R/
###  kernels.R
### Kernel functions for "Kernel adjusted density estimation" of Srihera &
### Stute (2011) and for "Rank Transformations in Kernel Density Estimation"
### of Eichner & Stute (2013)
### R 3.4.1, 13./14.2./28.9.2017 (21./24./26./27./28./31.10./4./9.11./5.12.2016)
###*****************************************************************************

#' Epanechnikov kernel
#'
#' Vectorized evaluation of the Epanechnikov kernel.
#'
#' @param x Numeric vector.
#' @return A numeric vector of the Epanechnikov kernel evaluated at the
#'         values in \code{x}.
#'
#' @examples
#' kader:::epanechnikov(x = c(-sqrt(6:5), -2:2, sqrt(5:6)))
#' \donttest{
#' curve(kader:::epanechnikov(x), from = -sqrt(6), to = sqrt(6))
#' }
#'
epanechnikov <- function(x) {
  (abs(x) < sqrt(5)) * 3 / (4 * sqrt(5)) * (1 - x*x / 5)
}


#' Rectangular kernel
#'
#' Vectorized evaluation of the rectangular kernel.
#'
#' @param x Numeric vector.
#' @param a Numeric scalar: lower bound of kernel support; defaults to -0.5.
#' @param b Numeric scalar: upper bound of kernel support; defaults to 0.5.
#' @return A numeric vector of the rectangular kernel evaluated at the
#'         values in \code{x}.
#' @examples
#' kader:::rectangular(x = seq(-1, 1, by = 0.1))
#' \donttest{
#' curve(kader:::rectangular(x), from = -1, to = 1)
#' }
#'
rectangular <- function(x, a = -0.5, b = 0.5) {
  (a < x & x < b) / (b - a)
}


R/Jfunctions.R
#2345678901234567890123456789012345678901234567890123456789012345678901234567890
### ~/PapersBooksReportsReviewsTheses/PapersSelbst/FestschriftWS70/kader/R/
###  Jfunctions.R
### Fcts. related to the optimal transformation (J) in "Rank Transformations
### in Kernel Density Estimation" of Eichner & Stute (2013), like the solution
### J to the isoperimetric problem.
### R 3.4.2, 10./12./13./16.2./21./22./24.8./27./28.9./4.10.2017
###  (27./31.10./16.11./5.12.2016)
###*****************************************************************************

#' Cube-root that retains its argument's sign
#'
#' Computes \eqn{(x_1^{1/3}, \ldots, x_n^{1/3})} with \eqn{x_i^{1/3}} being
#' negative if \eqn{x_i < 0}.
#'
#' @param x Numeric vector.
#' @return Vector of same length and mode as input.
#'
#' @examples
#' kader:::cuberoot(x = c(-27, -1, -0, 0, 1, 27))
#' \donttest{
#' curve(kader:::cuberoot(x), from = -27, to = 27)
#' }
#'
cuberoot <- function(x) {
  sign(x) * abs(x)^(1/3)
}



#' pc
#'
#' Coefficient \eqn{p_c} of eq. (15.15) in Eichner (2017).
#'
#' p_c = 1/5 * (3c^2 - 5) / (3 - c^2) * c^2.
#'
#' For further details see p. 297 f. in Eichner (2017) and/or Eichner & Stute
#' (2013).
#'
#' @export
#' @param cc Numeric vector.
#' @return Vector of same length and mode as \code{cc}.
#'
#' @note \eqn{p_c} should be undefined for \eqn{c = \sqrt 3}, but \code{pc}
#'       is here implemented to return \code{Inf} in each element of its
#'       return vector for which the corresponding element in \code{cc}
#'       contains R's value of \code{sqrt(3)}.
#'
#' @examples
#' \donttest{
#' c0 <- expression(sqrt(5/3))
#' c1 <- expression(sqrt(3) - 0.01)
#' cgrid <- seq(1.325, 1.7, by = 0.025)
#' cvals <- c(eval(c0), cgrid, eval(c1))
#'
#' plot(cvals, pc(cvals), xaxt = "n", xlab = "c", ylab = expression(p[c]))
#' axis(1, at = cvals, labels = c(c0, cgrid, c1), las = 2)
#' }
#'
pc <- function(cc) {
  cc2 <- cc*cc
  res <- (3*cc2 - 5) / (3 - cc2) * cc2 / 5
  issqrt3 <- sapply(cc, function(x) isTRUE(all.equal(x, sqrt(3))))
  res[issqrt3] <- Inf
  res
}



#' qc
#'
#' Coefficient \eqn{q_c(u)} of eq. (15.15) in Eichner (2017).
#'
#' \eqn{q_c(u) = 2/5 * c^5 / (3 - c^2) * (1 - 2 * u)}
#'
#' For further details see p. 297 f. in Eichner (2017) and/or Eichner & Stute
#' (2013).
#'
#' @export
#' @param u Numeric vector.
#' @param cc Numeric constant, defaults to \eqn{\sqrt(5/3)}.
#' @return Vector of same length and mode as \code{u}.
#'
#' @note \eqn{q_c(u)} should be undefined for \eqn{c = \sqrt 3}, but \code{qc}
#'       is here implemented to return \code{Inf * (1 - 2*u)} if \code{cc}
#'       contains R's value of \code{sqrt(3)}.
#'
#' @examples
#' \donttest{
#' u <- c(0, 1)   # seq(0, 1, by = 0.1)
#' c0 <- expression(sqrt(5/3))
#' c1 <- expression(sqrt(3) - 0.05)
#' cgrid <- seq(1.4, 1.6, by = 0.1)
#' cvals <- c(eval(c0), cgrid, eval(c1))
#'
#' Y <- sapply(cvals, function(cc, u) qc(u, cc = cc), u = u)
#' cols <- rainbow(ncol(Y), end = 9/12)
#' matplot(u, Y, type = "l", lty = "solid", col = cols,
#'   ylab = expression(q[c](u)))
#' abline(h = 0, lty = "dashed")
#' legend("topright", title = "c", legend = c(c0, cgrid, c1),
#'   lty = 1, col = cols, cex = 0.8)
#' }
#'
qc <- function(u, cc = sqrt(5/3)) {
  if(!is.numeric(cc) || length(cc) > 1)
    stop("cc must be a numeric vector of length 1!")
  cc2 <- cc*cc
  if(isTRUE(all.equal(3, cc2))) {
    return(Inf * (1 - 2*u))
  } else return(2/5 * cc*cc2*cc2 / (3 - cc2) * (1-2*u))
}



#' J1
#'
#' Eq. (15.16) in Eichner (2017) as a result of Cardano's formula.
#'
#' Using, for brevity's sake, \eqn{J_{1a}(u, c) := -q_c(u)} and
#' \eqn{J_{1b}(u, c) := J_{1a}(u, c)^2 + p_c^3}, the definition of
#' \eqn{J_1} reads:
#'
#' \eqn{J_1(u, c) :=   [J_{1a}(u, c) + \sqrt(J_{1b}(u, c))]^{1/3}
#'                   + [J_{1a}(u, c) - \sqrt(J_{1b}(u, c))]^{1/3}}.
#'
#' For implementation details of \eqn{q_c(u)} and \eqn{p_c} see
#' \code{\link{qc}} and \code{\link{pc}}, respectively.
#'
#' For further mathematical details see Eichner (2017) and/or Eichner
#' & Stute (2013).
#'
#' @export
#' @inheritParams qc
#' @return Vector of same length and mode as \code{u}.
#'
#' @note Eq. (15.16) in Eichner (2017), and hence \eqn{J_1(u, c)}, requires
#'       \eqn{c} to be in \eqn{[\sqrt(5/3), 3)}. If \code{cc} does
#'       not satisfy this requirement a warning (only) is issued.
#'
#' @seealso \code{\link{J_admissible}}.
#'
#' @examples
#' \donttest{
#' u <- seq(0, 1, by = 0.01)
#' c0 <- expression(sqrt(5/3))
#' c1 <- expression(sqrt(3) - 0.01)
#' cgrid <- c(1.35, seq(1.4, 1.7, by = 0.1))
#' cvals <- c(eval(c0), cgrid, eval(c1))
#'
#' Y <- sapply(cvals, function(cc, u) J1(u, cc = cc), u = u)
#' cols <- rainbow(ncol(Y), end = 9/12)
#' matplot(u, Y, type = "l", lty = "solid", col = cols,
#'   ylab = expression(J[1](u, c)))
#' abline(h = 0)
#' legend("topleft", title = "c", legend = c(c0, cgrid, c1),
#'   lty = 1, col = cols, cex = 0.8)
#' }
#'
J1 <- function(u, cc = sqrt(5/3)) {
  if(!is.numeric(cc) || length(cc) > 1)
    stop("cc must be a numeric vector of length 1!")
  if(cc >= sqrt(3) || cc < sqrt(5/3))
    warning( "cc should be in [sqrt(5/3), 3)!")

  J1a <- -qc(u, cc)
  pc0 <- pc(cc)
  J1b <- J1a*J1a + pc0*pc0*pc0
  cuberoot(J1a + sqrt(J1b)) + cuberoot(J1a - sqrt(J1b))
}



#' J2
#'
#' Eq. (20) in Eichner (2017) (based on "Bronstein's formula for k = 3")
#'
#' \eqn{J_2(u, c) = 2\sqrt(-p_c) * sin(1/3 * arcsin(q_c(u) / (-p_c)^{3/2}))}
#'
#' For implementation details of \eqn{q_c(u)} and \eqn{p_c} see
#' \code{\link{qc}} and \code{\link{pc}}, respectively.
#'
#' For further mathematical details see Eichner (2017) and/or Eichner &
#' Stute (2013).
#'
#' @export
#' @param u Numeric vector.
#' @param cc Numeric constant, defaults to \eqn{\sqrt 5}.
#' @return Vector of same length and mode as \code{u}.
#'
#' @note Eq. (20) in Eichner (2017), and hence \eqn{J_2(u, c)}, requires
#'       \eqn{c} to be in \eqn{(\sqrt 3, \sqrt 5]}. If \code{cc} does
#'       not satisfy this requirement (only) a warning is issued.
#'
#'       The default \code{cc = sqrt(5)} yields the optimal rank
#'       transformation.
#'
#' @seealso \code{\link{J_admissible}}.
#'
#' @examples
#' \donttest{
#' u <- seq(0, 1, by = 0.01)
#' c0 <- expression(sqrt(3) + 0.01)
#' c1 <- expression(sqrt(5))
#' cgrid <- seq(1.85, 2.15, by = 0.1)
#' cvals <- c(eval(c0), cgrid, eval(c1))
#'
#' Y <- sapply(cvals, function(cc, u) J2(u, cc = cc), u = u)
#' cols <- rainbow(ncol(Y), end = 9/12)
#' matplot(u, Y, type = "l", lty = "solid", col = cols,
#'   ylab = expression(J[2](u, c)))
#' abline(h = 0)
#' legend("topleft", title = "c", legend = c(c0, cgrid, c1),
#'   lty = 1, col = cols, cex = 0.8)
#' }
#'
J2 <- function(u, cc = sqrt(5)) {
  if(!is.numeric(cc) || length(cc) > 1)
    stop("cc must be a numeric vector of length 1!")
  if(cc <= sqrt(3) || cc > sqrt(5))
    warning( "cc should be in (sqrt(3), sqrt(5)]!")
  root <- sqrt(-pc(cc))
  qq <- qc(u, cc)
  asin.arg <- ifelse(qq == 0, 0, qq / (root*root*root))
              # To manage "0/0" correctly whenever cc = 0.
  asin.arg <- pmax(pmin(asin.arg, 1), -1) # To assure |asin.arg| <= 1.
  2 * root * sin(asin(asin.arg) / 3)
}



#' Admissible Rank Transformations of Eichner & Stute (2013)
#'
#' This is just a wrapper for the functions \code{\link{J1}}, \code{\link{J2}},
#' and \eqn{u -> \sqrt 3 * (2u - 1)} which implement the admissible
#' transformations for the three cases for \eqn{c}. For mathematical
#' details see eq. (15.16) and (15.17) in Eichner (2017) and/or Eichner &
#' Stute (2013).
#'
#' Basically, for \code{cc} in \eqn{[\sqrt(5/3), \sqrt 5]}:
#'
#' \code{J_admissible(u, cc)} = \code{J1(u, cc)} if \code{cc} \eqn{< \sqrt 3},
#'
#' \code{J_admissible(u, cc)} = \code{J2(u, cc)} if \code{cc} \eqn{> \sqrt 3},
#' and
#'
#' \code{J_admissible(u, cc)} = \code{sqrt(3) * (2*u - 1)} if \code{cc}
#' \eqn{= \sqrt 3}.
#'
#' @export
#' @inheritParams J2
#' @return Vector of same length and mode as \code{u}.
#'
#' @note The admissible rank transformations require \eqn{c} to
#'       be in \eqn{[\sqrt(5/3), \sqrt 5]}. If \code{cc} does
#'       not satisfy this requirement a warning (only) is issued.
#'       The default \code{cc = sqrt(5)}, i.e., \eqn{c = \sqrt 5},
#'       yields the optimal rank transformation.
#'
#' @seealso \code{\link{J1}} and \code{\link{J2}}.
#'
#' @examples
#' \donttest{
#' par(mfrow = c(1, 2), mar = c(3, 3, 0.5, 0.5), mgp = c(1.7, 0.7, 0))
#' example(J1)
#' example(J2)
#' }
#'
J_admissible <- function(u, cc = sqrt(5)) {
  if(!is.numeric(cc) || length(cc) > 1)
    stop("cc must be a numeric vector of length 1.")
  if(cc < sqrt(5/3) || cc > sqrt(5))
    stop("cc is out of the admissible range [sqrt(5/3), sqrt(5)]!")
  if(cc < sqrt(3)) {
    J1(u, cc)               # eq. (15.16) in Eichner (2017)
  } else if(cc > sqrt(3)) {
    J2(u, cc)               # eq. (15.17) in Eichner (2017)
  } else {   # cc = sqrt(3)
   sqrt(3) * (2*u - 1)
  }
}


R/minimizeMSE.R
#2345678901234567890123456789012345678901234567890123456789012345678901234567890
### ~/PapersBooksReportsReviewsTheses/PapersSelbst/FestschriftWS70/kader/R/
###  minimizeMSE.R
### Minimization of MSE-estimator for "Kernel adjusted density estimation" of
### Srihera & Stute (2011) and for "Rank Transformations in Kernel Density
### Estimation" of Eichner & Stute (2013).
### R 3.4.2, 13./14./15./16./17./24.2./22./23.8./28.9./2.10.2017
###  (21./24./26./27./28./31.10./4./9.11./5.12.2016)
###*****************************************************************************


#' Minimization of Estimated MSE
#'
#' Minimization of the estimated MSE as function of \eqn{\sigma} in four steps.
#'
#' Step 1: determine first (= smallest) maximizer of \code{VarHat.scaled} (!)
#' on the grid in \code{sigma}. Step 2: determine first (= smallest) minimizer
#' of estimated MSE on the \eqn{\sigma}-grid LEFT OF the first maximizer of
#' \code{VarHat.scaled}. Step 3: determine a range around the yet-found
#' (discrete) minimizer of estimated MSE within which a finer search for the
#' ``true'' minimum is continued using numerical minimization. Step 4: check if
#' the numerically determined minimum is indeed better, i.e., smaller than the
#' discrete one; if not keep the first.
#'
#' @inheritParams bias_AND_scaledvar
#' @param VarHat.scaled Vector of estimates of the scaled variance
#'                      (for values of \eqn{\sigma} in \code{sigma}).
#' @param BiasHat.squared Vector of estimates of the squared bias
#'                        (for values of \eqn{\sigma} in \code{sigma}).
#' @param plot Should graphical output be produced? Defaults to \code{FALSE}.
#' @param \ldots Currently ignored.
#'
#' @return A list with components \code{sigma.adap}, \code{msehat.min} and
#'         \code{discr.min.smaller} whose meanings are as follows:
#'   \tabular{ll}{
#'    \code{sigma.adap} \tab Found minimizer of MSE estimator. \cr
#'    \code{msehat.min} \tab Found minimum of MSE estimator. \cr
#'    \code{discr.min.smaller} \tab TRUE iff the numerically found minimum was
#'                             smaller than the discrete one.
#'    }
#'
#' @examples
#' require(stats)
#'
#' set.seed(2017);     n <- 100;     Xdata <- sort(rnorm(n))
#' x0 <- 1;      Sigma <- seq(0.01, 10, length = 11)
#'
#' h <- n^(-1/5)
#' Ai <- (x0 - Xdata)/h
#' fnx0 <- mean(dnorm(Ai)) / h   # Parzen-Rosenblatt estimator at x0.
#'
#'  # For non-robust method:
#' Bj <- mean(Xdata) - Xdata
#'#  # For rank transformation-based method (requires sorted data):
#'# Bj <- -J_admissible(1:n / n)   # rank trafo
#'
#' BV <- kader:::bias_AND_scaledvar(sigma = Sigma, Ai = Ai, Bj = Bj,
#'   h = h, K = dnorm, fnx = fnx0, ticker = TRUE)
#'
#' kader:::minimize_MSEHat(VarHat.scaled = BV$VarHat.scaled,
#'   BiasHat.squared = (BV$BiasHat)^2, sigma = Sigma, Ai = Ai, Bj = Bj,
#'   h = h, K = dnorm, fnx = fnx0, ticker = TRUE, plot = FALSE)
#'
minimize_MSEHat <- function(VarHat.scaled, BiasHat.squared, sigma, Ai, Bj, h,
  K, fnx, ticker = FALSE, plot = FALSE, ...) {

 message("Minimizing MSEHat:")

 # Step 1: determine first (= smallest) maximizer of VarHat.scaled (!!!)
 #         on the grid in sigma.
 #**********************************************************************
 message("Step 1: Search smallest maximizer of VarHat.scaled on sigma-grid.")
 varhatscaled.maxidx <- which.max(VarHat.scaled) # 1st sigma-index, where
                                                 # VarHat.scaled is maximal.

 if(plot) {
   graphics::points(sigma[varhatscaled.maxidx], # mark maximum in graph.
                    VarHat.scaled[varhatscaled.maxidx],
                    pch = 6, cex = 2, col = "violet", lwd = 2)
 }

 # Step 2: determine first (= smallest) minimizer of MSEHat on the sigma-grid
 #         LEFT OF the first maximizer of VarHat.scaled.
 #***************************************************************************
 message("Step 2: Search smallest minimizer of MSEHat on sigma-grid to the\n",
         "        LEFT of just found smallest maximizer of VarHat.scaled.")
 MSEHat <- BiasHat.squared + VarHat.scaled
 msehat.minidx <- which.min(MSEHat[ 1:varhatscaled.maxidx])
  # First sigma-index where MSEHat is minimal left of first
  # maximizer of VarHat.scaled.

 sigadap <- sigma[ msehat.minidx]
  # sigma-value for which MSEHat is minimal
  # left of first maximizer of VarHat.scaled.
 msehat.min <- MSEHat[ msehat.minidx]  # Pertaining minimal MSEHat-value


 # Step 3: determine a range around the yet-found (discrete) minimizer of
 #         MSEHat within which a finer search for the "true" minimum con-
 #         tinues using numerical minimization
 #***********************************************************************
 message("Step 3: Finer search for 'true' minimum of MSEHat using\n",
         "        numerical minimization. (May take a while.)")
 sigrange <- c(    sigma[max(      1,       msehat.minidx - 3)],
               min(sigma[min(length(sigma), msehat.minidx + 3)],
                   sigma[varhatscaled.maxidx]) )
 if(diff(sigrange) < .Machine$double.eps^0.25) sigrange <- range(sigma)

 if(plot) {
   graphics::segments(
     x0 = sigrange[1], y0 = graphics::par("usr")[3] / 2,   # mark range
     x1 = sigrange[2], col = "cyan1", lwd = 2)             # in graph
 }

 # Numerical minimization:
 # suppressMessages( {
 msehat.opt <- stats::optimize(mse_hat, interval = sigrange, Ai = Ai, Bj = Bj,
   h = h, K = K, fnx = fnx, ticker = ticker);  if(ticker) message("")
 # } )


 # Step 4: check if the numerically determined minimum is indeed better,
 #         i.e., smaller than the discrete one! And if not keep the first.
 #************************************************************************
 message("Step 4: Check if numerically determined minimum is smaller\n",
         "        than discrete one.")
 discrete.minimum.smaller.than.numerical <- FALSE
 if(msehat.opt$objective < msehat.min) {
   sigadap <- msehat.opt$minimum
   msehat.min <- msehat.opt$objective
   message("        Yes, optimize() was 'better' than grid search.")
 } else {
   message("        No, grid search was 'at least as good' as optimize().")
   discrete.minimum.smaller.than.numerical <- TRUE
   if(plot) {
     graphics::legend("bottom", legend = "Minimum adjusted!", bty = "n",
       cex = 1.3)
   }
 }


 if(plot) {
   graphics::points(c(msehat.opt$minimum, sigadap),       # mark "both min-
                    c(msehat.opt$objective, msehat.min),  # nima" in graph
                    pch = c(3, 4), cex = 2, col = c("green", "red"), lwd = 2)

   graphics::legend("topright", pch = c(6, NA, 3, 4), # cex = 0.8,
     lty = c(NA, "solid", NA, NA), lwd = 2, bty = "n", inset = c(0.005, 0),
     col = c("violet", "cyan1", "green", "red"),
     legend = c(
       expression("Found max. of"~sigma^2/(n*h^4) * widehat(Var)(sigma)),
       expression("Search-range for num. min. of"~widehat(MSE)(sigma)),
       substitute("Num. min." == group("(", list(x, y), ")"),
         list(x = round(msehat.opt$minimum, 4),
              y = round(msehat.opt$objective, 4))),
       substitute("Discr. min." == group("(", list(x, y), ")"),
         list(x = round(sigadap, 4),
              y = round(msehat.min, 4))) ) )
   }

 list(sigma.adap = sigadap, msehat.min = msehat.min,
      discr.min.smaller = discrete.minimum.smaller.than.numerical)
}
